hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/SqlEndToEndTestSupport.java

```diff
@@ -0,0 +1,155 @@
+    protected SqlServiceImpl sqlService;
+    protected PlanExecutor planExecutor;
+    protected JobCoordinationService jobCoordinationService;
+    protected PreJobInvocationObserverImpl preJobInvocationObserver;
+    protected JobInvocationObserverImpl jobInvocationObserver;
+
+
+    @Before
+    public void setUp() throws Exception {
+        nodeEngine = getNodeEngineImpl(instance());
+        preJobInvocationObserver = new PreJobInvocationObserverImpl();
+        jobInvocationObserver = new JobInvocationObserverImpl();
+        sqlService = (SqlServiceImpl) instance().getSql();
+        planExecutor = sqlService.getOptimizer().getPlanExecutor();
+        jobCoordinationService = ((JetServiceBackend) nodeEngine
+                .getService(JetServiceBackend.SERVICE_NAME))
+                .getJobCoordinationService();
+
+        planExecutor.registerJobInvocationObserver(preJobInvocationObserver);
+        jobCoordinationService.registerInvocationObserver(jobInvocationObserver);
+
+        eec = new ExpressionEvalContextImpl(
+                emptyList(),
+                Util.getSerializationService(instance()),
+                Util.getNodeEngine(instance()));
+    }
+
+    SqlPlanImpl.SelectPlan assertQueryPlan(String query) {
+        SqlStatement sql = new SqlStatement(query);
+        SqlPlan plan = sqlService.prepare(
+                sql.getSchema(),
+                query,
+                sql.getParameters(),
+                SqlExpectedResultType.ROWS
+        );
+
+        assertInstanceOf(SqlPlanImpl.SelectPlan.class, plan);
+        return (SqlPlanImpl.SelectPlan) plan;
+    }
+
+    void assertQueryResult(SqlPlanImpl.SelectPlan selectPlan, Collection<Row> expectedResults, Object... args) {
+        List<Object> arguments = Collections.emptyList();
+        if (args.length > 0) {
+            arguments = Arrays.asList(args);
+            eec = new ExpressionEvalContextImpl(
+                    arguments,
+                    Util.getSerializationService(instance()),
+                    Util.getNodeEngine(instance()));
+        }
+        QueryId queryId = QueryId.create(UUID.randomUUID());
+        SqlResult result = planExecutor.execute(selectPlan, queryId, arguments, 0L);
+        assertCollection(expectedResults, collectResult(result));
+    }
+
+    @Nonnull
+    protected static ArrayList<Row> collectResult(SqlResult result) {
+        var actualRows = new ArrayList<Row>();
+        for (SqlRow r : result) {
+            actualRows.add(new Row(r));
+        }
+        return actualRows;
+    }
+
+    protected static class PreJobInvocationObserverImpl implements PreJobInvocationObserver {
+        public DAG dag;
+        public JobConfig jobConfig;
+
+        @Override
+        public void onJobInvocation(DAG dag, JobConfig config) {
+            this.dag = dag;
+            this.jobConfig = config;
+        }
+    }
+
+    protected static class JobInvocationObserverImpl implements JobInvocationObserver {
```

Comment:
can't we use mockito mock+verify for the observers? wouldn't it be simpler and more flexible?

Reply:
> wouldn't it be simpler and more flexible?

You're reviewing the simpler way for me now :)